src/composer/prover.rs

```diff
@@ -0,0 +1,511 @@
+        let prover = Builder::prove(self.constraints, circuit)?;
+
+        let constraints = self.constraints;
+        let size = self.size;
+
+        let domain = EvaluationDomain::new(constraints)?;
+
+        let mut transcript = self.transcript.clone();
+
+        let public_inputs = prover.public_inputs();
+        let public_input_indexes = prover.public_input_indexes();
+        let dense_public_inputs = Builder::dense_public_inputs(
+            &public_input_indexes,
+            &public_inputs,
+            self.size,
+        );
+
+        public_inputs
+            .iter()
+            .for_each(|pi| transcript.append_scalar(b"pi", pi));
+
+        // round 1
+        // convert wires to padded scalars
+        let mut a_w_scalar = vec![BlsScalar::zero(); size];
+        let mut b_w_scalar = vec![BlsScalar::zero(); size];
+        let mut o_w_scalar = vec![BlsScalar::zero(); size];
+        let mut d_w_scalar = vec![BlsScalar::zero(); size];
+
+        prover.constraints.iter().enumerate().for_each(|(i, c)| {
+            a_w_scalar[i] = prover[c.w_a];
+            b_w_scalar[i] = prover[c.w_b];
+            o_w_scalar[i] = prover[c.w_o];
+            d_w_scalar[i] = prover[c.w_d];
+        });
+
+        let a_w_poly = Self::blind_poly(rng, &a_w_scalar, 1, &domain);
+        let b_w_poly = Self::blind_poly(rng, &b_w_scalar, 1, &domain);
+        let o_w_poly = Self::blind_poly(rng, &o_w_scalar, 1, &domain);
+        let d_w_poly = Self::blind_poly(rng, &d_w_scalar, 1, &domain);
+
+        // commit to wire polynomials
+        // ([a(x)]_1, [b(x)]_1, [c(x)]_1, [d(x)]_1)
+        let a_w_poly_commit = self.commit_key.commit(&a_w_poly)?;
+        let b_w_poly_commit = self.commit_key.commit(&b_w_poly)?;
+        let o_w_poly_commit = self.commit_key.commit(&o_w_poly)?;
+        let d_w_poly_commit = self.commit_key.commit(&d_w_poly)?;
+
+        // Add wire polynomial commitments to transcript
+        transcript.append_commitment(b"a_w", &a_w_poly_commit);
+        transcript.append_commitment(b"b_w", &b_w_poly_commit);
+        transcript.append_commitment(b"c_w", &o_w_poly_commit);
+        transcript.append_commitment(b"d_w", &d_w_poly_commit);
+
+        // round 2
+        // permutation challenges
+        let beta = transcript.challenge_scalar(b"beta");
+        transcript.append_scalar(b"beta", &beta);
+
+        let gamma = transcript.challenge_scalar(b"gamma");
+        let sigma = [
+            &self.prover_key.permutation.s_sigma_1.0,
+            &self.prover_key.permutation.s_sigma_2.0,
+            &self.prover_key.permutation.s_sigma_3.0,
+            &self.prover_key.permutation.s_sigma_4.0,
+        ];
+        let wires = [
+            a_w_scalar.as_slice(),
+            b_w_scalar.as_slice(),
+            o_w_scalar.as_slice(),
+            d_w_scalar.as_slice(),
+        ];
+        let permutation = prover
+            .perm
+            .compute_permutation_vec(&domain, wires, &beta, &gamma, sigma);
+
```

Comment:
For consistency, consider adding comment along the line:
// Commit to permutation polynomial and add to transcript

Reply:
Added!